### PR TITLE
add validation for spec.affinity

### PR DESCRIPTION
**Pull request by simonyangcj**
Signed-off-by: Simon Yang <yang.chenjun@99cloud.net>

**What this PR does / why we need it**:
So far vmi validation webhook does not validate spec.affinity fo vmi template

**Which issue(s) this PR fixes**
Fixes #<https://github.com/kubevirt/kubevirt/issues/8100>

**Special notes for your reviewer**:

**Release note**:
```release-note
Support validation for spec.affinity of vmi template
```

### CODE REVIEW

**Comment by rthallisey** [comment]
/ok-to-test
cc @alaypatel07 for review

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 1]
```diff
@@ -2473,3 +2482,236 @@ func virtioNicRequested(interfaces []v1.Interface) bool {
 
 	return false
 }
+
+// validateScheduler is function that validate spec.affinity
+// instead of bring in the whole kubernetes lib we simply copy it from kubernetes/pkg/apis/core/validation/validation.go
+// and vendor/k8s.io.apimachinery has to be updated because it's not updated to the latest version of the tag v0.23.5
+// which will leads to missing directory k8s.io/apimachinery/pkg/apis/meta/v1/validation eventually will cause we are unable
+// to copy validateAffinity function from kubernetes/pkg/apis/core/validation/validation.go
+func validateScheduler(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec) (causes []metav1.StatusCause) {
```
Comment (rmohr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:2491]: If we decide to go with copying, would you mind having all copied code in a dedicated file? https://github.com/kubevirt/kubevirt/blob/main/pkg/controller/expectations.go#L18 may be a good precedence.

If possible let's just keep the whole file pretty much unmodified to make detecting changes easy and future copying.

**Review by rmohr — COMMENTED** [review]
I found the dry-run approach also pretty nice. It definitely adds one extra api call per CRU operation. That should in most cases be fine, but if we can easily copy, it may be better to maybe have once or twice a compatibility regression and safe the calls ...

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 1 above)
Comment (simonyangcj) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:2491]: @rmohr will do, but I still having a hard time on failure of "pull-kubevirt-verify-go-mod" test. Looks like  update vendor/k8s.io/apimachinery break down the test.  I remove the dir in `vendor/k8s.io/apimachinery` and run `go mod vendor ` again to force a update and copy to vendor. Any ideal what did wrong here ? 

Thanks a lot :)

Comment (rmohr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:2491]: > Looks like update vendor/k8s.io/apimachinery break down the test. I remove the dir in vendor/k8s.io/apimachinery and run go mod vendor again to force a update and copy to vendor. Any ideal what did wrong here ?

Hm, I think you should just manually bump in `go.mod` and in the go.mod files in `staging` all references to 0.23.5 and then run `make dep-sync`.

Comment (simonyangcj) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:2483]: I'll give a try.  Thanks

**Comment by simonyangcj** [comment]
Hi @rmohr after a lot retry, I don't think I'm able solve the dependency issue, I think i'm going switch to dry run approach which won't touch any dependencies.  So I will close this PR and open up a new one.

**Comment by rmohr** [comment]
> Hi @rmohr after a lot retry, I don't think I'm able solve the dependency issue, I think i'm going switch to dry run approach which won't touch any dependencies. So I will close this PR and open up a new one.

Maybe it makes sense to go one step back and start with a clean branch without already copied code. I tried the bump in https://github.com/rmohr/kubevirt/tree/dep-sync-experiment (last two commits). First one just replaces all v0.23.1 with v0.23.5 in all relevant go.mod files, then I simply ran for the second commit `make deps-sync`. I did not encounter any issues.

**Comment by rmohr** [comment]
Note that we prune all unused dependencies from the vendor folder, so after copying over code afterwards, some imports may be missing from the vendor folder. Just run `make deps-sync` again (which will run go mod prune) to get them back.

**Comment by simonyangcj** [comment]
> Note that we prune all unused dependencies from the vendor folder, so after copying over code afterwards, some imports may be missing from the vendor folder. Just run `make deps-sync` again (which will run go mod prune) to get them back.

@rmohr Hi  I just create another pr with dry approach. Please took a look https://github.com/kubevirt/kubevirt/pull/8212

Thanks a lot:)

**Pull request closed without merging**